sky/skylet/log_lib.py

```diff
@@ -440,15 +476,39 @@ def tail_logs(job_id: Optional[int],
         with open(log_path, 'r', newline='', encoding='utf-8') as log_file:
             # Using `_follow` instead of `tail -f` to streaming the whole
             # log and creating a new process for tail.
+            if tail > 0:
+                current_file_lines = log_file.readlines()
+                lines = collections.deque(current_file_lines, maxlen=tail)
```

Comment:
One fruit of thought: I think the original purpose we use `deque` is to avoid the large memory consumption for a large log file, but here we are reading the whole file into the memory, so the memory consumption will still be high. Can we think of a way to avoid the large memory consumption by log files (we have users encountered the large memory consumption on controller)?

There are two ways:
1. Can we avoid reading the whole file when tail is provided or not?
2. If needed, we can also try to switch to `tail -f` / `tail -f -n` and find some way to handle the `start_stream_at`

Reply:
Yes, we use `deque` and iterable readlines() to save memory. 
I guess I should keep my overcomplicated implementation with cases 1, 2, and 3, with only a lookahead of the log file instead of consuming the whole file.